lib/Bindings/Python/circt/design_entry.py

```diff
@@ -55,55 +57,85 @@ def set(self, val):
+    _ODS_REGIONS = (0, True)
+
     def __init__(self, *args, **kwargs):
-      super().__init__(*args, **kwargs)
+      """Scan the class and eventually instance for Input/Output members and
+      treat the inputs as operands and outputs as results."""
+
+      # Copy the classmember Input/Output declarations to be instance members.
+      for attr_name in dir(cls):
+        attr = self.__getattribute__(attr_name)
+        if isinstance(attr, Input):
+          self.__setattr__(attr_name, Input(attr.type, attr_name))
+        if isinstance(attr, Output):
+          self.__setattr__(attr_name, Output(attr.type, attr_name))
+
+      cls.__init__(self, *args, **kwargs)
+
+      # The OpView attributes cannot be touched before OpView is constructed.
+      # Get a list and don't touch them.
+      dont_touch = set([x for x in dir(mlir.ir.OpView)])
 
       # After the wrapped class' construct, all the IO should be known.
-      input_ports = []
-      output_ports = []
+      self.input_ports = list[Input]()
+      self.output_ports = list[Output]()
+      # Scan for them.
       for attr_name in dir(self):
+        if attr_name in dont_touch:
+          continue
         attr = self.__getattribute__(attr_name)
-        if type(attr) is Input:
-          input_ports.append((attr_name, attr))
-        if type(attr) is Output:
-          output_ports.append((attr_name, attr))
-
-      # This function sets up the inputs to construct, then connects the
-      # outputs.
-      def body_build(mod):
-        inputs = dict()
-        for index, (name, _) in enumerate(input_ports):
-          inputs[name] = mod.entry_block.arguments[index]
-
-        self.construct(**inputs)
-
-        outputs = []
-        unconnected_ports = []
-        for (name, output) in output_ports:
-          if output.value is None:
-            unconnected_ports.append(name)
-          outputs.append(output.value)
-        if len(unconnected_ports) > 0:
-          raise UnconnectedSignalError(cls.__name__, unconnected_ports)
-        hw.OutputOp(outputs)
-
-      # Construct things as HWModules.
-      self.module = hw.HWModuleOp(
-          name=cls.__name__,
-          input_ports=[(name, port.type) for (name, port) in input_ports],
-          output_ports=[(name, port.type) for (name, port) in output_ports],
-          body_builder=body_build)
+        if isinstance(attr, Input):
+          attr.name = attr_name
+          self.input_ports.append(attr)
+        if isinstance(attr, Output):
+          attr.name = attr_name
+          self.output_ports.append(attr)
+
+      # Replace each declared input with an MLIR value so we look similar to the
+      # other OpViews.
+      input_ports_values = list[mlir.ir.Value]()
+      for input in self.input_ports:
+        if input.name in kwargs:
+          value = kwargs[input.name]
+          if isinstance(value, mlir.ir.OpView):
+            value = value.operation.result
+          elif isinstance(value, mlir.ir.Operation):
```

Comment:
This is the same idea as the logic of the connect API. I'm wondering if the Input and Output here can inherit from BuilderValue and we can put this logic there. I was thinking about this earlier before I saw this logic.